src/models/instance.ts

```diff
@@ -119,40 +120,42 @@ export default class Instance {
         return new Instance({ ...this.data, description });
     }
 
-    public encryptPassword(encryptionKey: string) {
-        const cryptr = new Cryptr(encryptionKey);
-        const encrypted = cryptr.encrypt(this.data.password);
-        return new Instance({ ...this.data, password: encrypted });
+    public async encryptPassword(): Promise<Instance> {
+        const encryptionKey = await Instance.getEncryptionKey();
+        const password =
+            this.data.password.length > 0
+                ? new Cryptr(encryptionKey).encrypt(this.data.password)
+                : "";
+        return new Instance({ ...this.data, password });
     }
 
-    public decryptPassword(encryptionKey: string) {
-        const cryptr = new Cryptr(encryptionKey);
-        const decrypted = cryptr.decrypt(this.data.password);
-        return new Instance({ ...this.data, password: decrypted });
+    public async decryptPassword(): Promise<Instance> {
+        const encryptionKey = await Instance.getEncryptionKey();
+        const password =
+            this.data.password.length > 0
+                ? new Cryptr(encryptionKey).decrypt(this.data.password)
+                : "";
+        return new Instance({ ...this.data, password });
     }
 
     public get description(): string {
         return this.data.description ? this.data.description : "";
     }
 
-    public async validateUrlUsernameCombo(d2: D2) {
+    public async validateUrlUsernameCombo(d2: D2): Promise<boolean> {
         const { url, username, id } = this.data;
         const combination = [url, username].join("-");
         const instanceArray = await getData(d2, instancesDataStoreKey);
         const invalidCombinations = instanceArray.filter(
             (inst: Data) => [inst.url, inst.username].join("-") === combination
         );
-        let invalid;
-        if (id) {
-            invalid = invalidCombinations.some((inst: Data) => inst.id !== id);
-        } else {
-            invalid = !_.isEmpty(invalidCombinations);
-        }
 
-        return invalid;
+        return id
+            ? invalidCombinations.some((inst: Data) => inst.id !== id)
+            : !_.isEmpty(invalidCombinations);
     }
 
-    public async validate(d2: D2) {
+    public async validate(d2: D2): Promise<any> {
```

Comment:
Okay, updated the PR with a solution based-off your comment. It's not that restrictive as you propose, any key will be valid instead of limiting to "name", "url", ... This makes sense if we want to re-use the validation types for other models other than Instance.

How does it look?

Reply:
Looks good to me!